docs/source/reference/types.rst

```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
    >>> numba.float32[::1, :, :]
    array(float32, 3d, F)
 
+Functions
+---------
+
+.. warning::
+   The feature of considering functions as first-class type objects is
+   under development.
+
+Functions are often considered as certain transformations of
+input arguments to output values. Within Numba :term:`JIT` compiled
+functions, the functions can also be considered as objects, that is,
+functions can be passed around as arguments or return values, or used
+as items in sequences, in addition to being callable.
+
+First-class function support is enabled for all Numba :term:`JIT`
+compiled functions and Numba ``cfunc`` compiled functions except when:
+- using a non-CPU compiler,
+- the compiled function is a Python generator,
+- the compiled function has Omitted arguments,
+- or the compiled function returns Optional value.
+
+To disable first-class function support, use ``no_cfunc_wrapper=True``
+decorator option.
+
+For instance, consider an example where the Numba :term:`JIT` compiled
+function applies user-specified functions as a composition to an input
+argument::
+
+    >>> @numba.njit
+    ... def composition(funcs, x):
+    ...     r = x
+    ...     for f in funcs[::-1]:
+    ...         r = f(r)
+    ...     return r
+    ...
+    >>> @numba.cfunc("double(double)")
+    ... def a(x):
+    ...     return x + 1.0
+    ...
+    >>> @numba.njit
+    ... def b(x):
+    ...     return x * x
+    ...
+    >>> composition((a, b), 0.5), 0.5 ** 2 + 1
+    (1.25, 1.25)
+    >>> composition((b, a, b, b, a), 0.5), b(a(b(b(a(0.5)))))
+    (36.75390625, 36.75390625)
+
+Here, ``cfunc`` compiled functions ``a`` and ``b`` are considered as
+first-class function objects because these are passed in to the Numba
+:term:`JIT` compiled function ``composition`` as arguments, that is, the
+``composition`` is :term:`JIT` compiled independently from its argument function
+objects (that are collected in the input argument ``funcs``).
+
+Currently, first-class function objects can be Numba ``cfunc`` compiled
+functions, :term:`JIT` compiled functions, and objects that implement the
+Wrapper Address Protocol (WAP, see below) with the following restrictions:
+
+========================   ==============   ============    ===========
+Context                    cfunc compiled   jit compiled    WAP objects
+========================   ==============   ============    ===========
+Can be used as arguments   yes              yes             yes
+Namespace scoping          yes              yes             no
+Can be called              yes              yes             yes
+Can be used as items       yes              yes             yes
+Automatic overload         no               yes             no
+Can be returned            yes              yes             yes
+========================   ==============   ============    ===========
```

Comment:
nit: consider sorting the `no`s together for easier reading
```suggestion
========================   ============   ==============   ===========
Context                    jit compiled   cfunc compiled   WAP objects
========================   ============   ==============   ===========
Can be used as arguments   yes            yes              yes
Can be called              yes            yes              yes
Can be used as items       yes            yes              yes
Can be returned            yes            yes              yes
Namespace scoping          yes            yes              no
Automatic overload         yes            no               no
========================   ============   ==============   ===========
```
(swapped rows and columns)

Reply:
Good idea.

The table needs the following update as well:
`Namespace scoping` for `WAP` is `yes`.